tests/fuser_mqubit_test.cc

```diff
@@ -644,6 +648,333 @@ TEST(FuserMultiQubitTest, SmallCircuits) {
   }
 }
 
+TEST(FuserMultiQubitTest, ValidTimeOrder) {
+  using Fuser = MultiQubitGateFuser<IO, DummyGate>;
+
+  Fuser::Parameter param;
+  param.verbosity = 0;
+
+  {
+    unsigned num_qubits = 8;
+    std::vector<DummyGate> circuit = {
+      CreateDummyGate(0, {0, 1}),
+      CreateDummyGate(0, {2, 3}),
+      CreateDummyControlledGate(1, {1}, {2}),
+      CreateDummyGate(0, {4, 5}),
+      CreateDummyGate(2, {0, 1}),
+      CreateDummyGate(1, {3, 4}),
+      CreateDummyGate(2, {2, 3}),
+      CreateDummyGate(3, {1, 2}),
+      CreateDummyControlledGate(2, {4}, {5}),
+      CreateDummyGate(3, {3, 4}),
+      CreateDummyGate(5, {0, 1}),
+      CreateDummyGate(4, {2, 3}),
+      CreateDummyGate(5, {4, 5}),
+      CreateDummyGate(4, {6, 7}),
+    };
+
+    param.max_fused_size = 2;
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end());
+
+    EXPECT_EQ(fused_gates.size(), 14);
+    EXPECT_TRUE(TestFusedGates(num_qubits, circuit, fused_gates));
+  }
+
+  {
+    unsigned num_qubits = 6;
+    std::vector<DummyGate> circuit = {
+      CreateDummyGate(0, {0, 1}),
+      CreateDummyGate(0, {2, 3}),
+      CreateDummyGate(1, {1, 2}),
+      CreateDummyGate(0, {4, 5}),
+      CreateDummyGate(1, {3, 4}),
+      CreateDummyMeasurementGate(2, {0, 1, 2}),
+      CreateDummyMeasurementGate(2, {4, 5}),
+      CreateDummyGate(3, {0, 1}),
+      CreateDummyGate(3, {2, 3}),
+      CreateDummyGate(4, {1, 2}),
+      CreateDummyGate(3, {4, 5}),
+      CreateDummyGate(4, {3, 4}),
+    };
+
+    param.max_fused_size = 6;
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end());
+
+    EXPECT_EQ(fused_gates.size(), 3);
+    EXPECT_TRUE(TestFusedGates(num_qubits, circuit, fused_gates));
+  }
+
+  {
+    unsigned num_qubits = 6;
+    std::vector<DummyGate> circuit = {
+      CreateDummyGate(0, {0, 1}),
+      CreateDummyGate(0, {2, 3}),
+      CreateDummyGate(1, {1, 2}),
+      CreateDummyGate(0, {4, 5}),
+      CreateDummyGate(1, {3, 4}),
+      CreateDummyGate(1, {5, 0}),
+    };
+
+    param.max_fused_size = 3;
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end());
+
+    EXPECT_EQ(fused_gates.size(), 4);
```

Comment:
Which gates fuse in this example? I guess the reason it's 4 instead of 3 is that fusing the last pair would form a loop - i.e. `(4,5) -> (0,1)-(1,2) -> (2,3)-(3,4) -> (5,0)`?

Reply:
Exactly, the last pair would form a loop and we have to avoid that.